geode-core/src/distributedTest/java/org/apache/geode/internal/cache/RebalanceWhileCreatingRegionDistributedTest.java

```diff
@@ -0,0 +1,298 @@
+
+  @Rule
+  public ClusterStartupRule cluster = new ClusterStartupRule();
+
+  @Rule
+  public SerializableTestName testName = new SerializableTestName();
+
+  @Rule
+  public DistributedBlackboard blackboard = new DistributedBlackboard();
+
+  private static final Logger logger = LogService.getLogger();
+
+  public static final String BEFORE_REMOVE_BUCKET_MESSAGE = "Before_RemoveBucketMessage";
+
+  public static final String AFTER_CREATE_PROXY_REGION = "After_CreateProxyRegion";
+
+  @Test
+  public void testRebalanceDuringRegionCreation() throws Exception {
+    // Init Blackboard
+    blackboard.initBlackboard();
+
+    // Start Locator
+    MemberVM locator = cluster.startLocatorVM(0);
+
+    // Start servers
+    int locatorPort = locator.getPort();
+    MemberVM server1 = cluster.startServerVM(1, locatorPort);
+    MemberVM server2 = cluster.startServerVM(2, locatorPort);
+    MemberVM accessor = cluster.startServerVM(4, locatorPort);
+
+    // Add DistributionMessageObserver
+    String regionName = testName.getMethodName();
+    Stream.of(server1, server2, accessor)
+        .forEach(server -> server.invoke(() -> addDistributionMessageObserver(regionName)));
+
+    // Create regions in each server
+    server1.invoke(() -> createRegion(regionName, RegionShortcut.PARTITION));
+    server2.invoke(() -> createRegion(regionName, RegionShortcut.PARTITION));
+
+    // Asynchronously wait to create the proxy region in the accessor
+    accessor.invokeAsync(() -> waitToCreateProxyRegion(regionName));
+
+    // Connect client1
+    ClientVM client1 =
+        cluster.startClientVM(5, c -> c.withServerConnection(server1.getPort(), server2.getPort()));
+
+    // Do puts
+    client1.invoke(() -> {
+      Region<Integer, Integer> region =
+          ClusterStartupRule.clientCacheRule.createProxyRegion(regionName);
+      IntStream.range(0, 3).forEach(i -> region.put(i, i));
+    });
+
+    // Start server3
+    MemberVM server3 = cluster.startServerVM(3, locatorPort);
+
+    // Create region in server3
+    server3.invoke(() -> createRegion(regionName, RegionShortcut.PARTITION));
+
+    // Add DistributionMessageObserver to server3
+    server3.invoke(() -> addDistributionMessageObserver(regionName));
+
+    // Rebalance
+    server1.invoke(() -> ClusterStartupRule.getCache().getResourceManager().createRebalanceFactory()
+        .start().getResults());
+
+    // Stop server3
+    server3.invoke(() -> ClusterStartupRule.getCache().close());
+
+    // Connect client to accessor
+    ClientVM client2 =
+        cluster.startClientVM(6, c -> c.withServerConnection(accessor.getPort())
+            .withCacheSetup(cf -> cf.setPoolReadTimeout(20000)));
+
+    // Do puts
+    client2.invoke(() -> {
+      Region<Integer, Integer> region =
+          ClusterStartupRule.clientCacheRule.createProxyRegion(regionName);
+      IntStream.range(0, 3).forEach(i -> region.put(i, i));
+    });
```

Comment:
It will be good to add validation for some of the expected state. 
- Wait for accessor to finish and see it proxy-region creation completed successfully. Accessing that region in accessor node to fetch and perform cache operation (say put and get) to see there is no issue with the region and all the expected data is present.

Reply:
Additional verification added.